proposals/simd-0170-builtin-instruction-cost-and-budget.md

```diff
@@ -0,0 +1,126 @@
+---
+simd: '0170'
+title: Specifying CU Definitions for Builtins
+authors:
+  - Tao Zhu (Anza)
+category: Standard
+type: Core
+status: Review
+created: 2024-08-26
+feature: https://github.com/anza-xyz/agave/issues/2562
+supersedes: 
+superseded-by:
+extends:
+---
+
+## Summary
+
+1. Builtin programs should consume a predefined number of CUs for each
+   instruction.
+2. Since builtins can invoke other programs (CPI), they should allocate enough
+   but granular CUs to ensure successful execution without over-allocation.
+
+## Motivation
+
+This proposal addresses two key issues related to CU allocation and consumption
+for builtin programs while aiming to avoid adding unnecessary complexity.
+
+1. Accurate CU tracking without post-execution adjust-up: Currently,
+   builtin instructions deduct a fixed amount of CUs (DEFAULT_COMPUTE_UNITS)
+from both the CU meter and block limits after execution. However, the CU meter
+allocates a much larger value (DEFAULT_INSTRUCTION_COMPUTE_UNIT_LIMIT) to each
+builtin instruction. This discrepancy creates imprecise CU tracking and may
+require block producers to account for additional CUs after a transaction's
+execution, potentially causing block limits to be exceeded.
+
+   Furthermore, builtins can conditionally CPI into other programs, introducing
+variability in CU consumption. This unpredictability makes it difficult to
+allocate CUs upfront solely based on CUs builtin consumes during execution, also
+adds complexity to the tracking process.
+
+2. Preventing over-allocation of CUs: Over-allocating CUs for builtins
+   reduces block density, lowers network throughput, and can degrade block
+producer performance by causing repeated transaction retries. Avoiding excessive
+CU allocation is critical for maximizing block efficiency and minimizing network
+delays.
+
+To resolve these issues, we propose statically defining the CU cost of builtin
+instructions based on worst-case scenarios, including potential CPI calls. A
+unified MAX_BUILTIN_ALLOCATION_COMPUTE_UNIT_LIMIT will standardize CU allocation
+for both block producers and the CU meter.
+
+## New Terminology
+
+None
+
+## Detailed Design
+
+1. Retain the existing DEFAULT_COMPUTE_UNITS for builtin programs.
+   When the virtual machine (VM) invokes a builtin instruction, the
+statically defined DEFAULT_COMPUTE_UNITS is consistently deducted from the CU
+Meter.
```

Comment:
Let's leave this out of this SIMD because it's existing behavior which isn't changing. I think if firedancer is interested in how this existing behavior works, we can detail it in the follow up SIMD that @tao-stones will be writing for per-ix cu costs.